src/OpenTelemetry.Api/Trace/Tracer.cs

```diff
@@ -17,64 +17,174 @@
 using System;
 using System.Collections.Generic;
 using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
 
 namespace OpenTelemetry.Trace
 {
     /// <summary>
     /// Tracer to record distributed tracing information.
     /// </summary>
-    public abstract class Tracer
+    public class Tracer
     {
+        internal readonly ActivitySource ActivitySource;
+
+        internal Tracer(ActivitySource activitySource)
+        {
+            this.ActivitySource = activitySource;
+        }
+
         /// <summary>
         /// Gets the current span from the context.
         /// </summary>
-        public abstract TelemetrySpan CurrentSpan { get; }
+        public TelemetrySpan CurrentSpan
+        {
+            get
+            {
+                return new TelemetrySpan(Activity.Current);
+            }
+        }
 
         /// <summary>
-        /// Activates the span on the current context.
+        /// Starts span.
         /// </summary>
-        /// <param name="span">Span to associate with the current context.</param>
-        /// <param name="endSpanOnDispose">Flag indicating if span should end when scope is disposed.</param>
-        /// <returns>Disposable object to control span to current context association.</returns>
-        public abstract IDisposable WithSpan(TelemetrySpan span, bool endSpanOnDispose);
+        /// <param name="name">Span name.</param>
+        /// <param name="kind">Kind.</param>
+        /// <returns>Span instance.</returns>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
```

Comment:
I'm just curious about `AggressiveInlining` added throughout this PR. Besides knowing that it exists, I've never used it before. Having worked on auto-instrumenting product before, I have had reason to use `NoInlining`, but not `AggressiveInlining`. What's the intuition for when it's appropriate?

Reply:
It is a micro optimization for wrapper methods or helper methods. https://www.dotnetperls.com/aggressiveinlining

The general guideline is to measure performance before start tuning 😄